elastic/elastic-agent
# [Upgrade Details] Ensure details report `UPG_WATCHING` for the entire time that the upgrade is being watched

## What does this PR do?

This PR fixes the upgrade details such that they are in the `UPG_WATCHING` state the entire time the Agent upgrade is being watched by the Upgrade Watcher.

## Why is it important?

So the state of the upgrade is accurately reported.

## Checklist

- [ ] ~I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~ Bug was never released.

## How to test this PR locally

1. Build Elastic Agent from this PR but give it a lower-than-current version number.  This will be the starting (pre-upgrade) version of the Agent.
   ```
   AGENT_PACKAGE_VERSION=8.11.0 EXTERNAL=true SNAPSHOT=true PLATFORMS=darwin/arm64 PACKAGES=targz mage package
   ```
2. Make a no-op commit.  Without this the Agent upgrade will not succeed.
3. Build Elastic Agent again. This will be the target (post-upgrade) version of the Agent.
   ```
   EXTERNAL=true SNAPSHOT=true PLATFORMS=darwin/arm64 PACKAGES=targz mage package
   ```
4. Install the starting version of the Agent.
5. Upgrade to the target version of the Agent.
   ```
   sudo elastic-agent upgrade 8.12.0-SNAPSHOT --source-uri file:///Users/shaunak/development/github/elastic-agent/build/distributions --skip-verify
   ```
6. Check that the Agent status reports upgrade details with a state of `UPG_WATCHING`.
   ```
   sudo elastic-agent status --output json | jq '.upgrade_details'
   {
     "target_version": "8.12.0",
     "state": "UPG_WATCHING",
     "metadata": {}
   }
   ```
8. Check the Agent logs and verify the upgrade details states are in order.  In particular, make sure that **chronologically**, we see `UPG_WATCHING` after `UPG_RESTARTING`.
   ```
   sudo grep -R -h --include=\*.ndjson UPG_ /Library/Elastic/Agent | jq -c -s 'sort_by(.["@timestamp"]) | .[]'
   {"log.level":"info","@timestamp":"2023-11-28T11:13:45.530Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":499},"message":"updated upgrade details","log":{"source":"elastic-agent"},"upgrade_details":{"target_version":"8.12.0-SNAPSHOT","state":"UPG_REQUESTED","metadata":{}},"ecs.version":"1.6.0"}
   {"log.level":"info","@timestamp":"2023-11-28T11:13:45.531Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":499},"message":"updated upgrade details","log":{"source":"elastic-agent"},"upgrade_details":{"target_version":"8.12.0-SNAPSHOT","state":"UPG_DOWNLOADING","metadata":{}},"ecs.version":"1.6.0"}
   {"log.level":"info","@timestamp":"2023-11-28T11:13:45.729Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":499},"message":"updated upgrade details","log":{"source":"elastic-agent"},"upgrade_details":{"target_version":"8.12.0-SNAPSHOT","state":"UPG_EXTRACTING","metadata":{}},"ecs.version":"1.6.0"}
   {"log.level":"info","@timestamp":"2023-11-28T11:13:52.308Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":499},"message":"updated upgrade details","log":{"source":"elastic-agent"},"upgrade_details":{"target_version":"8.12.0-SNAPSHOT","state":"UPG_REPLACING","metadata":{}},"ecs.version":"1.6.0"}
   {"log.level":"info","@timestamp":"2023-11-28T11:13:52.313Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":499},"message":"updated upgrade details","log":{"source":"elastic-agent"},"upgrade_details":{"target_version":"8.12.0-SNAPSHOT","state":"UPG_REPLACING","metadata":{}},"ecs.version":"1.6.0"}
   {"log.level":"info","@timestamp":"2023-11-28T11:13:52.320Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":499},"message":"updated upgrade details","log":{"source":"elastic-agent"},"upgrade_details":{"target_version":"8.12.0-SNAPSHOT","state":"UPG_RESTARTING","metadata":{}},"ecs.version":"1.6.0"}
   {"log.level":"info","@timestamp":"2023-11-28T11:14:00.374Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":499},"message":"updated upgrade details","log":{"source":"elastic-agent"},"upgrade_details":{"target_version":"8.12.0","state":"UPG_WATCHING","metadata":{}},"ecs.version":"1.6.0"}
   ```
6. Wait until the Upgrade Watcher has finished running.
   ```
   pgrep -f 'elastic-agent watch' | wc -l    # should report 0 eventually
   ```
9. Check the Agent version and verify that it's the target version.
   ```
   sudo elastic-agent version

   Binary: 8.12.0-SNAPSHOT (build: 80bb6a61369c20e054478a73c6c866aadfcc52b1 at 2023-11-27 23:38:37 +0000 UTC)
   Daemon: 8.12.0-SNAPSHOT (build: 80bb6a61369c20e054478a73c6c866aadfcc52b1 at 2023-11-27 23:38:37 +0000 UTC)
   ```
11. Cleanup: revert/remove the no-op commit from step 2.

## Related issues
- Fixes #3816
